### DISCUSSION TITLE
Using the loading icon for visual feedback

### DISCUSSION THREAD

**Discussion by fdinhSG** [Kitware/trame#254] [Q&A] [answered]
Since I have a process that takes several minutes I want to block a button with a loading icon. I saw from [here](https://vuetifyjs.com/en/components/buttons/) that one could use the `loading` prop in a button. But I cannot make it work.

This is a sample code:

``` python
from trame.app import get_server
from trame.ui.vuetify import VAppLayout
from trame.widgets import vuetify
import time

server = get_server()
state, ctrl = server.state, server.controller

async def wait_a_bit():
    state.loading = True
    time.sleep(3)
    print('Hello!')
    state.loading = False

with VAppLayout(server) as layout:
    vuetify.VBtn("Hi", click=lambda: wait_a_bit(), loading=('loading', False))

server.start()
```

I'd expect the button to transform into the loading icon for 3 seconds and then to transform back, but it doesn't change at all. I've tried defining the button from the start with `loading=('loading', True)` and that works, so there must be something wrong in the function call. How can I do that?

Thanks!

**Comment by jourdain** [comment]
Ah the trickiness of Python async. ;-)

Let me explain.

First of all the full `wait_a_bit` method will prevent the server for processing any other concurrent request. It can be fine, but you may want to schedule task internally if you want to "free" the server sooner.

Next the state get sync after each method call, that is why the client never see the switch. 

To force the flush of the state at some given time you should do the following:

```python

async def wait_a_bit():
    with state:
       state.loading = True
    time.sleep(3)
    print('Hello!')
    with state: # Might not be needed but good practice to do it in async flow
       state.loading = False
```

**Reply by fdinhSG** [reply]
Thank you for your reply!

However, it still looks like it isn't working. When I click on the button, 3 seconds pass, then for the briefest moment a loading logo appears and it transforms back into "HI". It's barely noticeable.

Can you try the following script and tell me if it works on your machine?

``` python
from trame.app import get_server
from trame.ui.vuetify import VAppLayout
from trame.widgets import vuetify
import time

server = get_server()
state, ctrl = server.state, server.controller

async def wait_a_bit():
    with state:
       state.loading = True
    time.sleep(3)
    print('Hello!')
    with state: # Might not be needed but good practice to do it in async flow
       state.loading = False

with VAppLayout(server) as layout:
    vuetify.VBtn("Hi", click=lambda: wait_a_bit(), loading=('loading', False))

server.start()
```

**Reply by jourdain** [reply]
Yes I realized afterward that your sleep was not an `async.sleep` but a busy/locking one... ;-)

**Comment by jourdain** [comment, marked as answer]
Ideally, I would write your code that way

```python
import asyncio
from trame.app import get_server, asynchronous
from trame.ui.vuetify import VAppLayout
from trame.widgets import vuetify
import time

server = get_server()
state, ctrl = server.state, server.controller

@asynchronous.task
async def wait_a_bit():
    with state:
        state.loading = True
    await asyncio.sleep(0.01) # Give room for server to handle network
    time.sleep(3) # Busy 
    with state:
        print('Hello!')
        state.loading = False

with VAppLayout(server) as layout:
    vuetify.VBtn("Hi", click=wait_a_bit, loading=('loading', False))

server.start()
```

**Reply by fdinhSG** [reply]
Ah, I missed the asynchronous.task decorator! async tasks are still tricky to me. Thanks!

**Reply by jourdain** [reply]
They are... I still get bitten from time to time... ;-)
